streams/src/main/java/org/apache/kafka/streams/kstream/Suppressed.java

```diff
@@ -0,0 +1,88 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream;
+
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressedImpl;
+import org.apache.kafka.streams.kstream.internals.suppress.SuppressedImpl.IntermediateSuppressionImpl;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+import java.time.Duration;
+
+@SuppressWarnings("unused") // complaint about the top-level generic parameters, which actually are necessary
+public interface Suppressed<K, V> {
+    interface TimeDefinition<K, V> {
+        long time(ProcessorContext context, K k, V v);
+    }
+
+    interface EagerBufferConfig<K, V> extends BufferConfig<K, V, EagerBufferConfig<K, V>> {
+
+    }
+
+    interface StrictBufferConfig<K, V> extends BufferConfig<K, V, StrictBufferConfig<K, V>> {
+
+    }
+
+    interface BufferConfig<K, V, BC extends BufferConfig<K, V, BC>> {
+        static <K, V> EagerBufferConfig<K, V> boundedByKeys(final long maxKeys) {
+            return new SuppressedImpl.EagerBCImpl<>(maxKeys);
+        }
+
+        BC bufferKeys(final long maxKeys);
+
+        static <K, V> EagerBufferConfig<K, V> boundedBySize(final long maxBytes) {
+            return new SuppressedImpl.EagerBCImpl<>(maxBytes);
+        }
+
+        BC bufferBytes(final long maxBytes);
+
+        static <K, V> StrictBufferConfig<K, V> unbounded() {
+            return new SuppressedImpl.StrictBCImpl<>();
+        }
+
+        StrictBufferConfig<K, V> unboundedBuffer();
+
+        StrictBufferConfig<K, V> spillToDiskWhenFull();
+
+        StrictBufferConfig<K, V> shutDownWhenFull();
+
+        EagerBufferConfig<K, V> emitEarlyWhenFull();
+    }
+
+    interface IntermediateSuppression<K, V> {
+        static <K, V> IntermediateSuppression<K, V> withEmitAfter(final Duration timeToWaitForMoreEvents) {
+            final IntermediateSuppression<K, V> intermediateSuppression = new IntermediateSuppressionImpl<>();
+            return intermediateSuppression.emitAfter(timeToWaitForMoreEvents);
+        }
+
+        IntermediateSuppression<K, V> emitAfter(final Duration timeToWaitForMoreEvents);
+
+        static <K, V> IntermediateSuppression<K, V> withBufferConfig(final BufferConfig<K, V, ? extends BufferConfig> bufferConfig) {
+            final IntermediateSuppression<K, V> intermediateSuppression = new IntermediateSuppressionImpl<>();
+            return intermediateSuppression.bufferConfig(bufferConfig);
+        }
+
+        IntermediateSuppression<K, V> bufferConfig(final BufferConfig<K, V, ? extends BufferConfig> bufferConfig);
+    }
+
+    static <K extends Windowed, V> Suppressed<K, V> emitFinalResultsOnly(final StrictBufferConfig<K, V> bufferConfig) {
+        return new SuppressedImpl<>(bufferConfig);
+    }
+
+    static <K, V> Suppressed<K, V> intermediateEvents(final IntermediateSuppression<K, V> intermediateSuppression) {
```

Comment:
Yeah, this is another spot where I really struggled with naming. I liked `suppress(intermediateEvents())` aka "suppress intermediate events", but the version for final would be like "suppress all but final events", and `suppress(allButFinalEvents()` just seems too confusing, so I compromised.

Note that this method isn't saying to _emit_ the events, but actually the opposite: to _suppress_ them. I think the symmetric name would be `suppressIntermediateEvents()`, which looks a little redundant in practice: `suppress(suppressIntermediateEvents())` or `suppress(Suppressed.suppressIntermediateEvents())` :/

Another idea would be to choose one of the synonyms. Since we are accomplishing this suppression via buffering, we could call it `bufferIntermediateEvents` or just `buffer`...

Thoughts?

Reply:
I see. And with this reasoning I think I also like `intermediateEvents` as well.

How about `suppress(intermediateEvents())` and `suppress(untilWindowEnds())` since the latter should be only called for windowed table result, and hence putting the keyword `window` as part of the func name should be fine?